python/foglamp/services/core/api/filters.py

```diff
@@ -0,0 +1,113 @@
+from foglamp.common.storage_client.exceptions import StorageServerError
+from foglamp.common import utils
+from foglamp.services.core.api import utils as apiutils
+from foglamp.common import logger
+
+__author__ = "Massimiliano Pinto"
+__copyright__ = "Copyright (c) 2018 OSIsoft, LLC"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+_help = """
+    -----------------------------------------------------------------------
+    | POST            | /foglamp/filter                               |
+    -----------------------------------------------------------------------
+"""
+
+_LOGGER = logger.setup("filter")
+
+async def add_filter(request):
+    """
+    Create a new filter with a specific plugin
+    
+    :Example:
+     curl -X POST http://localhost:8081/foglamp/filter -d 
+     '{
+        "name": "North_Readings_to_PI_scale_stage_1Filter",
+        "plugin": "scale"
+     }'
+
+    'name' is the filter name
+    'plugin' is the filter plugin name
+
+    The plugin is loaded and default comnfg from 'plugin_info'
+    is fetched.
+
+    A new config category 'name' is created:
+    items are:
+       - 'plugin'
+       - all ityems from default plugin config
+
+    NOTE: The 'create_category' call is made with keep_original_items = True
+
+    """
+    try:
+        data = await request.json()
+        filter_name = data.get('name')
+        plugin_name = data.get('plugin')
+        filter_desc = 'Configuration of \'' + filter_name + '\' filter for plugin \'' + plugin_name + '\'' 
+
+        # Get configuration manager instance
+        cf_mgr = ConfigurationManager(connect.get_storage_async())
+
+        # Load the specified plugin and get plugin data
+        loaded_plugin_info = apiutils.get_plugin_info(plugin_name)
+
+        # Get plugin default configuration (dict)
+        plugin_config = loaded_plugin_info['config']
+
+        # Get plugin type (string)
+        loaded_plugin_type = loaded_plugin_info['type'];
+
+        # Get plugin name (string)
+        loaded_plugin_name = plugin_config['plugin']['default'];
+
+        # Sanity checks
+        if (plugin_name != loaded_plugin_name or loaded_plugin_type != 'filter'):
+             errorMessage = "Loaded plugin '" + loaded_plugin_name + \
+                            "', type '" + loaded_plugin_type + \
+                            "', doesn't match the specified one '" + \
+                            plugin_name + "', type 'filter'"
+             raise Exception(errorMessage)
+
+        #################################################
+        # Set string value for 'default' if type is JSON
+        # This is required by the configuration manager
+        ################################################# 
+        for key, value in plugin_config.items():
+            if (value['type'] == 'JSON'):
+                value['default'] = json.dumps(value['default'])
+
```

Comment:
We should check if the category already exists as we are adding a new filter here explicitly, so the category should not already exist. This differs from the add code for service and task as it is not an error to create a category that already exists.

Reply:
Fixed: it filter already exists an error is returned:

400: Filter 'Scale10Filter' already exist.